### PR TITLE
Create Route finished, along with code coverage over threshold

### PR DESCRIPTION
**Pull request by tanaytrivedi**
Made my create route, resolving merge issues as I make this.

### CODE REVIEW

File: tests/test_routes.py [hunk 1]
```diff
@@ -23,6 +25,7 @@ class TestYourResourceServer(TestCase):
     @classmethod
```
Comment (loixe) [inline tests/test_routes.py:25]: line 22 TestPromotionServer(TestCase)

File: tests/test_routes.py [hunk 2]
```diff
@@ -46,27 +49,81 @@ def test_index(self):
         self.assertEqual(len(resp.get_json()), 6) #test number of endpoints coming back
     
     def test_create(self):
-        """ Calls create endpoint. """
-        resp = self.client.post("/promotions")
-        self.assertEqual(resp.status_code, status.HTTP_200_OK)
+        """ It should respond to a proper create with 201 status code and return the data. """
+        promo = PromoFactory()
+        data_orig = promo.serialize()
+        del data_orig['id'] # user is not supposed to send ID, they're supposed to receive it
```
Comment (loixe) [inline tests/test_routes.py:55]: Doesn't quite make sense to me. if you delete the id, other tests can't get it by id.

File: tests/test_routes.py [hunk 3]
```diff
@@ -46,27 +49,81 @@ def test_index(self):
...
+            headers={
+                'Content-type':'application/json', 
+                'Accept':'application/json'
+            }
+        )
+        self.assertEqual(resp.status_code, status.HTTP_201_CREATED)
+        response_json = resp.get_json()
+        del response_json['id']
+        del response_json['original_end_date']
+        del data['original_end_date']
```
Comment (loixe) [inline tests/test_routes.py:69]: And here too

File: tests/test_routes.py [hunk 4]
```diff
@@ -10,8 +10,10 @@
 from unittest import TestCase
 from unittest.mock import MagicMock, patch
 from service import app
-from service.models import db
+from service.models import Promotion, DataValidationError, db
 from service.common import status  # HTTP Status Codes
+from service.helpers import convert_data, convert_data_back
+from tests.factories import PromoFactory
```
Comment (loixe) [inline tests/test_routes.py:17]: why don't we have a “_create_promotions” function here?

Comment (loixe) [reply tests/test_routes.py:17]: It would be great if you can add some logging.debug in tests?

**Review by loixe — COMMENTED** [review]
It would be great if you can add some logging.debug in tests

File: tests/test_routes.py (hunk 2 above)
Comment (tanaytrivedi) [reply tests/test_routes.py:55]: So when ben made the factory he made it with all the fields that a Promo would have in the database. Thats for backend testing

But the frontend API is made for the users to utilize. Users dont set their own ID. the ID is returned to them by the endpoint for them to keep track of or to recover from the list endpoint. 

If you're setting up tests for delete/update/read, please run a create first in your test function and then don't delete the ID AFTER the create so you can use it to test your endpoint.

File: tests/test_routes.py (hunk 3 above)
Comment (tanaytrivedi) [reply tests/test_routes.py:69]: See my reply above.

**Comment by tanaytrivedi** [comment]
Added some more logging and brought code coverage up a bit. Cannot bring code coverage to 95% until someone does the update endpoint and its tests.

**Comment by tanaytrivedi** [comment]
Leaving PR open until midnight est tonight for more comments. Then I will merge it.

File: tests/test_routes.py (hunk 4 above)
Comment (loixe) [reply tests/test_routes.py:18]: _create_promotions function will create promotions in bulk which would be helpful and convenient for reading and listing promotions.

**Review by loixe — COMMENTED** [review]
Could you please add some logging in test_routes as well?

File: service/helpers.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+from datetime import date
+import datetime
+from service.models import DataValidationError
+
+from . import app
+
+def convert_data(data):
```
Comment (fzyyy0601) [inline service/helpers.py:7]: Hi, I saw this code, and from my understanding you are converting the date between two forms (The "yyyy-mm-dd" and timestamp). But I didn't see any place that timestamp exists. Is there any specific reasons that you add these two functions here?

**Review by Harshbansal1407 — APPROVED** [review]

File: service/helpers.py (hunk 5 above)
Comment (tanaytrivedi) [reply service/helpers.py:7]: The dates are datetime objects within the Promotion object. they come as strings through the REST API. Thus need something to go back and forth. 

Bools are similar.

File: tests/test_routes.py (hunk 4 above)
Comment (tanaytrivedi) [reply tests/test_routes.py:18]: Its a good idea, I am open to writing it for a future use case. I am adding an issue in the Icebox in Zenhub for us to make this in the future.

**Comment by tanaytrivedi** [comment]
Final code coverage at 93%. Again, can't improve it on my own, update and its tests must bridge the gap.

**Pull request merged**